### ISSUE TITLE
OOM Occurred on Single RTX 4090

### ISSUE THREAD

**Issue by tonybare** [ErgastiAlex/MARS#3]
Hello, thank you for your outstanding research.

In your paper, you mentioned that the experiments were conducted on an RTX 4090 GPU. I am using the same GPU, though I only have a single card. I followed the package versions and the launch script you provided.

However, when using a batch size of 8, I encountered an issue on the RSTPReid dataset where the VRAM usage gradually increased with each iteration, eventually leading to an OOM (Out of Memory) error. On the CUHK-PEDES dataset, the OOM error occurred right at the first iteration.

Could you kindly advise if I might have overlooked something? Also, could you please let me know how many RTX 4090 GPUs were used in total for your experiments? Below, I’ve listed the package versions that I believe might be relevant:
```
Package                           Version
--------------------------------- ------------------
torch                            1.9.1+cu111
torchvision                      0.10.1+cu111
transformers                     4.8.1
timm                             0.4.9
spacy                            3.7.4
spacy-legacy                     3.0.12
spacy-loggers                    1.0.5
einops                           0.6.1
```
And my launch script is:
```
export CUDA_VISIBLE_DEVICES=0
#export CUDA_VISIBLE_DEVICES=4,5,6,7

python -m torch.distributed.run --nproc_per_node=1 \
Retrieval.py \
--config configs/PS_rstp_reid.yaml \
--output_dir output/rstp-reid/train \
--checkpoint /data/MARS/ALBEF.pth \
--eval_mAP
```
Thank you very much for your time!

**Comment by ErgastiAlex** [comment]
Hi,
I trained on a single 4090 gpu. 
This is the conda enviroment that i used:
```
Package                  Version
------------------------ -----------
annotated-types          0.7.0
blis                     1.0.1
catalogue                2.0.10
certifi                  2024.8.30
charset-normalizer       3.3.2
click                    8.1.7
cloudpathlib             0.19.0
confection               0.1.5
contourpy                1.3.0
cycler                   0.12.1
cymem                    2.0.8
einops                   0.8.0
en_core_web_sm           3.8.0
filelock                 3.16.1
fonttools                4.54.1
fsspec                   2024.9.0
huggingface-hub          0.0.12
idna                     3.10
importlib_resources      6.4.5
Jinja2                   3.1.4
joblib                   1.4.2
kiwisolver               1.4.7
langcodes                3.4.1
language_data            1.2.0
marisa-trie              1.2.0
markdown-it-py           3.0.0
MarkupSafe               2.1.5
matplotlib               3.9.2
mdurl                    0.1.2
mpmath                   1.3.0
murmurhash               1.0.10
networkx                 3.2.1
numpy                    2.0.2
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-nccl-cu12         2.20.5
nvidia-nvjitlink-cu12    12.6.77
nvidia-nvtx-cu12         12.1.105
packaging                24.1
pillow                   10.4.0
pip                      24.2
preshed                  3.0.9
pydantic                 2.9.2
pydantic_core            2.23.4
Pygments                 2.18.0
pyparsing                3.1.4
python-dateutil          2.9.0.post0
PyYAML                   6.0.2
regex                    2024.9.11
requests                 2.32.3
rich                     13.9.1
ruamel.yaml.clib         0.2.8
ruamel-yaml-conda        0.17.21
sacremoses               0.1.1
setuptools               75.1.0
shellingham              1.5.4
six                      1.16.0
smart-open               7.0.4
spacy                    3.8.2
spacy-legacy             3.0.12
spacy-loggers            1.0.5
srsly                    2.4.8
sympy                    1.13.3
thinc                    8.3.2
timm                     0.4.9
tokenizers               0.10.3
torch                    2.4.1
torchaudio               2.4.1
torchvision              0.19.1
tqdm                     4.66.5
transformers             4.8.1
triton                   3.0.0
typer                    0.12.5
typing_extensions        4.12.2
urllib3                  2.2.3
wasabi                   1.1.3
weasel                   0.4.1
wheel                    0.44.0
wrapt                    1.16.0
zipp                     3.20.2
```

Probably instead of using torch 1.9.1 is better to use a recent version. I will update the readme accordingly as soon as possible.

To create the env you can just execute
```bash
conda create -n MARS python==3.9
conda activate MARS

pip3 install torch torchvision torchaudio
pip install transformers==4.8.1
pip install timm==0.4.9
pip install spacy
python -m spacy download en
pip install einops
conda install anaconda::ruamel_yaml
pip install matplotlib
```

Alex

**Comment by tonybare** [comment]
Unfortunately, after setting up the environment according to your configuration, I still encountered an OOM issue on the RSTPReid dataset after iteration 3600. However, the training on the CUHK-PEDES dataset was able to complete successfully without any issues. I’m wondering if there are any other areas I can further optimize or improve. Could you provide any suggestions?

Thank you very much for your time!

**Comment by ErgastiAlex** [comment]
I don't think that CUDA 12.1 is the problem. 
However, there are 2 quick fix that you can implement by yourself that will help you train the model with lower VRAM (even if i don't really know why on RSTPReid isn't working).
You can implement gradient accumulation or gradient checkpointing. The former you accumulate the gradient for multiple batches (e.g. so you can reduce the batch size by a factor of 2 and accumulate for 2 iteration). The latter is probabily the easier one to implement and is just a memory saving trick.

**Comment by MengXingYuu** [comment]
> CUHK-PEDES dataset

Hello, I have encountered the same issue. I am using an RTX 3090 GPU, and during the validation phase on the CUHK-PEDES dataset, an OOM (Out of Memory) error occurred. Have you resolved this problem?
![Snipaste_2024-11-26_14-14-45](https://github.com/user-attachments/assets/c602e67d-3241-4d80-900d-634f82b3b2de)

**Comment by ErgastiAlex** [comment]
Hi, you could reduce the test batch size inside the config file to avoid OOM errors during evaluation. However, if your GPU has 24GB of VRAM, it is quite unusual to encounter an OOM problem, as I am able to run the model seamlessly on my 4090.

Alex

**Comment by MengXingYuu** [comment]
> Hi, you could reduce the test batch size inside the config file to avoid OOM errors during evaluation. However, if your GPU has 24GB of VRAM, it is quite unusual to encounter an OOM problem, as I am able to run the model seamlessly on my 4090.
> 
> Alex

The issue has been resolved, sorry for bothering you. It was caused by a problem with my dataset.

**Comment by ErgastiAlex** [comment]
No problem!
Glad you solved!

**Comment by LGRlgr6666** [comment]
I want to ask, can a 4049 work
